### PR TITLE
This type predicates for type guards

### LINKED ISSUE #5764: Add support for "this is X" style user-defined type guards
I'd like to use user-defined type guards in the following manner:
```typescript
abstract class Animal
{
    get isDog(): this is Dog;
}

class Dog extends Animal
{
    get isDog(): this is Dog { return true; }
    bark() { console.log("woof"); }
}

let animal = new Dog();
if (animal.isDog)
{
   animal.bark()
}

```

This use-case was suggested by @RyanCavanaugh when he opened #1007 but seems to be unsupported as of TypeScript 1.6.

### PR DESCRIPTION
**Pull request by weswigham**
Closes #5764.

This type predicates are allowed anywhere both a type predicate and a this type are allowed, and additionally are allowed on class and interface _members_, allowing for this-predicate getters and this-predicate members.

### CODE REVIEW

**Comment by mhegazy** [comment]
nice :trophy:

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -1632,6 +1632,10 @@
         "category": "Error",
         "code": 2517
     },
+    "A this based type guard is not assignable to a parameter based type guard": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: 'this'-based

parameter-based

File: tests/cases/conformance/expressions/typeGuards/typeGuardFunctionOfFormThis.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
+class RoyalGuard {
```
Comment (DanielRosenwasser) [inline tests/cases/conformance/expressions/typeGuards/typeGuardFunctionOfFormThis.ts]: Can you add a `// @declaration: true`?

File: tests/cases/conformance/expressions/typeGuards/typeGuardFunctionOfFormThisErrors.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+class RoyalGuard {
```
Comment (DanielRosenwasser) [inline tests/cases/conformance/expressions/typeGuards/typeGuardFunctionOfFormThisErrors.ts]: // @declaration: true` here too

**Comment by DanielRosenwasser** [comment]
Can you add an example where you orphan your type guard method to a function?

File: src/compiler/types.ts [hunk 4]
```diff
@@ -1820,10 +1824,25 @@ namespace ts {
...
+        type: Type;
+    }
+
+    // @kind (TypePredicateKind.This)
+    export interface ThisTypePredicate extends TypePredicate {
+        _thisTypePredicateBrand: any;
+    }
+
+    // @kind (TypePredicateKind.Identifier)
+    export interface IdentifierTypePredicate extends TypePredicate {
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: It seems like overkill to have two different kinds of TypePredicate types and a dedicated enum. Could you just use `parameterIndex = -1` to indicate `this`?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I think a lot of this special casing could be eliminated by using `parameterName = "this"` and `parameterIndex = -1` to indicate a `this` type predicate.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -6691,19 +6746,35 @@ namespace ts {
...
-                if (signature.typePredicate &&
-                    expr.arguments[signature.typePredicate.parameterIndex] &&
-                    getSymbolAtLocation(expr.arguments[signature.typePredicate.parameterIndex]) === symbol) {
+                if (!signature.typePredicate) {
+                    return type;
+                }
+                const predicate = signature.typePredicate;
+                if (isIdentifierTypePredicate(predicate)) {
+                    if (expr.arguments[predicate.parameterIndex] &&
+                        getSymbolAtLocation(expr.arguments[predicate.parameterIndex]) === symbol) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: But the `getSymbolAtLocation` function is intended only for the language service and has fuzzy semantics that aren't part of the language. It's definitely not the right function to use here. I know this code is not part of the PR, but we should fix it.

File: src/compiler/checker.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I would really prefer we didn't use a magic number because you'd be special-casing in a slightly different but more implicit way.

Comment (weswigham) [inline src/compiler/checker.ts]: I disagree - I'd need the same special casing (except in clone), except the switch would be on `if (predicate.parameterName === "this" && predicate.parameterIndex === 1)` which is wholly uglier. This-based type guards and identifier-based types guards get parsed differently, checked differently, and find what to narrow differently - they actually share very little implementation, and the safety afforded by splitting it into two types (and then using a union of the two as the signature member, rather than the base type) is very nice. I do need to fix `cloneTypePredicate` to use the type guard I made though - I hadn't yet made it when I was writing it and neglected to go back and update it.

Additionally, over the past few hours I've realized that `this`-based type guards need to be extended to be allowed as property values/getter return types (this was in the original proposal and seems very reasonable), further separating them from identifier-based predicates.

The main thing they share is the slot on `Signature` which stores them, as they're mutually exclusive, but the way they're handled at every step is very different.

Comment (ahejlsberg) [inline src/compiler/checker.ts]: You wouldn't need to check the parameter name, you could just check `parameterIndex === -1`. I don't think the differences are as extensive as you make them out to be, the check would be no more expensive than what you have now, and I think you could eliminate a fair bit of code. Also, the AST uses a single type for the two forms, so why not do the same in the checker.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -5548,33 +5577,55 @@ namespace ts {
...
-                            const sourceTypeText = typeToString(source.typePredicate.type);
-                            const targetTypeText = typeToString(target.typePredicate.type);
-
-                            if (hasDifferentParameterIndex) {
-                                reportError(Diagnostics.Parameter_0_is_not_in_the_same_position_as_parameter_1,
-                                    sourceParamText,
-                                    targetParamText);
-                            }
-                            else if (hasDifferentTypes) {
+                            reportError(Diagnostics.A_this_based_type_guard_is_not_assignable_to_a_parameter_based_type_guard);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This special casing of this based type guards doesn't really seem necessary to me. I would just treat it like a parameter named 'this' and use the existing error messages.

File: src/compiler/checker.ts (hunk 5 above)
Comment (weswigham) [inline src/compiler/checker.ts]: The AST has a single root type (`TypePredicateNode`) with _two_ possible types for the `parameterName` member - `Identifier` or `ThisTypeNode` (could just as well be `ThisKeyword`, but a `ThisTypeNode` seems more appropriate). The current `TypePredicate` hierarchy reflects that - the base type can be either form, with `IdentifierTypePredicate` being specialized to the `Identifier` case and `ThisTypePredicate` being specialized for the `ThisTypeNode` case, since `TypePredicate` is structured via inheritance rather than composition (and refactoring it to match the grammar and be structured by composition would change a bunch code and be harder to adapt to by API consumers).

And they _are_ that different - in every case where we handle a type predicate in the checker (all two of them) there's a switch delegating to different behaviors for each kind.

Comment (ahejlsberg) [inline src/compiler/checker.ts]: I count multiple places (at least three) where you wouldn't need to branch if you did what I propose. In cases where you do branch, the code would be no larger. So, I conclude there is no doubt the code would be smaller if you follow my suggestion. I will leave it at that.

File: src/compiler/checker.ts (hunk 7 above)
Comment (weswigham) [inline src/compiler/checker.ts]: That wouldn't work well at all. The existing error messages look for matching parameter names when comparing two type guards - `Parameter "this" is not in the same position as parameter "x"` is a non-intuitive error, given that there is no parameter named "this" (or I would have reused it). They also don't check that the this-based type guard is in a location with a valid this-type.

**Comment by DanielRosenwasser** [comment]
@JsonFreeman I realized that the way that @weswigham has implemented this, we could discriminate on generators being finished or not.

**Comment by JsonFreeman** [comment]
@DanielRosenwasser Is that the case? If I understand you correctly, you mean that to check if an iterator/generator is done, you would do
```ts
if (iterResult.done) {
    ...
}
```
Is it possible to use a type predicate on a property declaration? If not, it looks like you'd have to define the `done` property as a getter. I'm not sure if this is a problem, but it's worth noting. I guess accessors are treated the same as properties, so in theory it might be okay, it just looks weird to define it that way.

The real issue is that a type predicate can only tell you *whether* `this` is of a certain type, given its done value. It cannot tell discriminate between *two* types based on the done value. So would the definition of IteratorResult be something like this (given a type SuspendedType for done: false, and CompletedType for done: true)
```ts
{
    value: SuspendedType;
    done: this is { value: CompletedType };
}
```
Is this the correct definition? I guess this is how you would discriminate between the two types. But if CompletedType is not assignable to SuspendedType, then will the type guard even apply? I think it will still give you SuspendedType.

**Comment by DanielRosenwasser** [comment]
> If I understand you correctly, you mean that to check if an iterator/generator is done, you would do
```ts
if (iterResult.done) {
    ...
}
```

@JsonFreeman yes, that's exactly correct.

@weswigham has allowed these predicate types on properties because it is necessary to support them on getters as well. It *is* slightly surprising (at least it was for me), but I don't necessarily see anything wrong with it<sup>[1]</sup>. It's no more than a `boolean` that is used as an indicator for the type, and that's being encoded into the type.

You might be correct in that weakness though @weswigham can weigh in. We should have some samples on this.

I think the interface we'd end up with is

```ts
interface IteratorResult<CompletedType, SuspendedType> {
    value: CompletedType | SuspendedType;
    done: this is { value: CompletedType };
}
```

<sup>[1]</sup> There's an issue regarding "orphaning" the property, whereby we don't orphan the type to `boolean`:

```ts
let x = {
    p: this is Whargarbl,
}
let y = x.p; // 'y' has type 'this is Whargarbl'
```

**Comment by JsonFreeman** [comment]
Oh... why is the type of `y` in your last example a type predicate? In the initial design of type predicates, they denoted type boolean, and the narrowing semantics did not flow around along with the value. I'm curious why this has changed.

Barring that, I think supporting type predicates on properties is fine.

As for IteratorResult, the type you wrote is technically accurate (more so than mine), but it spoils the element type of an iterator, which is supposed to be SuspendedType in the example. This was one of the major challenges of typing generators, and may have to be revisited if generators become more fully supported in the type system.

**Comment by DanielRosenwasser** [comment]
Is it spoiled? I think we could still appropriately negate the guard on the union of `value` and end up with `SuspendedType` like we would want - not `CompletedType | SuspendedType`.

**Comment by JsonFreeman** [comment]
That would involve type subtraction semantics. You would have to have way in the type system of seeing `{ value: CompletedType | SuspendedType }`, and then subtracting from it, `{ value: CompletedType }`, and getting the result `{ value: SuspendedType }`. In order to do this, you would need to make a lot of very specific assumptions about the structure of IteratorResult. That might be okay if you check that all these assumptions hold, when you process lib.d.ts. That way you could ensure up front that this type subtraction would be well defined when you needed it. Also, is the expectation that this subtraction would apply in the false branch of a type guard on `done`, in addition to iteration constructs?

**Comment by DanielRosenwasser** [comment]
But I believe we actually do "the right thing":

```ts
function f(x): x is { p: string } {
    return true;
}

var v: {
    p: string | number;
}

if (f(v)) {
    v.p // 'p' has type 'string'!
}
```

**Comment by JsonFreeman** [comment]
I mean in the else block.

**Comment by DanielRosenwasser** [comment]
Ah, that wouldn't be covered unfortunately, and you're right.

**Comment by JsonFreeman** [comment]
Why is a type predicate still effective after a property is orphaned?

**Comment by DanielRosenwasser** [comment]
Assignable

> Why is a type predicate still effective after a property is orphaned?

Because that's the way it just currently works. We were looking into how to shed it properly following an assignment. You can't do it as part of widening because you don't want to "dive in" to the type you're widening.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -4973,6 +5031,7 @@ namespace ts {
                 if (relation === assignableRelation) {
                     if (isTypeAny(source)) return Ternary.True;
                     if (source === numberType && target.flags & TypeFlags.Enum) return Ternary.True;
+                    if (source.flags & TypeFlags.Boolean && target.flags & TypeFlags.Boolean && (source.flags & TypeFlags.PredicateType || target.flags & TypeFlags.PredicateType)) return Ternary.True;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Consider

```ts
if (source.flags & target.flags & TypeFlags.Boolean && (source.flags | target.flags) & TypeFlags.Predicate)

**Comment by JsonFreeman** [comment]
Why not just have the property be of type `boolean` to begin with? The type predicate only needs to take effect on a property access type guard. And that can be achieved by fetching the type predicate associated with the property symbol. No types needed.

File: src/compiler/types.ts [hunk 9]
```diff
@@ -2089,6 +2108,7 @@ namespace ts {
         ESSymbol                = 0x01000000,  // Type of symbol primitive introduced in ES6
         ThisType                = 0x02000000,  // This type
         ObjectLiteralPatternWithComputedProperties = 0x04000000,  // Object literal type implied by binding pattern has computed properties
+        PredicateType           = 0x08000000,
```
Comment (JsonFreeman) [inline src/compiler/types.ts]: Why does a type predicate act as a type?

File: src/compiler/checker.ts (hunk 8 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I actually think the current way is clearer.

File: src/compiler/types.ts (hunk 9 above)
Comment (weswigham) [inline src/compiler/types.ts]: Why special case type predicates everywhere with extra slots for them when they're mutually exclusive with a return type (if there is a type predicate, the return type _must_ be boolean)? They even affect type equality (differently than booleans do) when they're nested within a type. Why not unify them _all_ as predicate types? It would be fairly easy to unify them in this way, IMO.

Yes, it being a type is why it follows assignments at present (like all types), but it shouldn't be hard to fix - it's just a matter of making the right branches of `getWidenedType` return `boolean` for type predicates.

The more interesting issue I've noted with them right now is that `this` types aren't properly reinstantiated when narrowing. I can't nest two `this` type guards, like so:
```ts
class FSO {
  isFile: this is File;
  isNetworked: this is (this & Networked);
}
class File {
  path: string;
}
interface Networked {
  host: string;
}
let f: FSO = new File();
if (f.isFile) {
  // f should be File
  if (f.isNetworked) {
    // f should be File & Networked, but is just File 
  }
}
```

Comment (JsonFreeman) [inline src/compiler/types.ts]: Hmm, that is an interesting issue with the nested 'this' type predicates. It does seem like it should work.

The original reason why type predicates were special cased, was to prevent them from flowing around. That could be taken care of with widening, but it still requires type relations for type predicates. The main thing is that the a type predicate is in intimately tied to one of the parameters in its associated signature (or `this`), so once it departs from the signature or property, it is really just a boolean. Is widening really enough to make sure it doesn't travel around?

Comment (weswigham) [inline src/compiler/types.ts]: It certainly seems like it, given that we widen on assignment.

Comment (JsonFreeman) [inline src/compiler/types.ts]: Yeah, I am thinking about it. You may be right actually. It's possible we did not think of widening at the time. @mhegazy or @tinganho might have thoughts on this point of whether to treat type predicates as types.

File: src/compiler/checker.ts (hunk 8 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I'm curious, why do you require at least one of the types to be a predicate type? Surely two plain old booleans are mutually assignable.

**Comment by weswigham** [comment]
On `this` type predicate members:

Technically, it is only "correct" to write a `this` type guard which checks for subtype assignability of the `this` type are predicates of the form:
```ts
interface I {
  member: this is (this & J);
} 
interface J {
  member2: string;
}
```
as conceivably, only a type containing this object's this type can be a subtype of the this type (This is also the only kind of type guard which does not become vacuously false for certain subclasses). _However_ this isn't want users want - users want to be able to narrow from base classes to sub-classes:
```ts
interface I {
  member: this is J
}
interface J extends I {
  member2: string;
}
```
therefore I propose not checking subtype assignability to the `this` type of this type guards (however, narrowing will still check if the type if a subtype to actually narrow). This means for certain subclasses a `this` type guard can be vacuously true or vacuously false (we have the information needed to know which and could model it for reachability checks if we add boolean value types).

File: src/compiler/checker.ts (hunk 8 above)
Comment (weswigham) [inline src/compiler/checker.ts]: Previously, all boolean compatibility was covered via reference comparison of the types. Now that there can be multiple boolean types, I needed to actually check for their assignability. However, while claiming that all boolean based types are assignable is correct for now, it would be incorrect once boolean value types are added.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Oh, so after adding the value types, if you have a set containing the types `boolean`, `true`, `false`, and type predicates, which pairs would you want to exclude from the assignability relation?

Comment (weswigham) [inline src/compiler/checker.ts]: Probably just true and false, from each other, honestly? Meh, you're probably right, it probably doesn't need to check for type-guardiness.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: One interesting question is whether `boolean` or a type predicate should be assignable to a value type like `true`. Intuitively, it should not be.

Comment (weswigham) [inline src/compiler/checker.ts]: I imagine not, since their truthiness is unknown.

File: src/compiler/types.ts (hunk 9 above)
Comment (weswigham) [inline src/compiler/types.ts]: Fixed the issue - it was simply that `this` type guards weren't re-instantiated by `instantiateType`. :+1:

**Comment by weswigham** [comment]
@JsonFreeman I've been able to unify _all_ type predicates into predicate types with minimal changes to baselines (our type baseline writer doesn't get widened types, so when it prints out the values of predicate returning functions it now prints the predicate rather than `boolean`, which is a mostly cosmetic change as far as our baselines are concerned) - the `typePredicate` field on `Signature` is no more. At @DanielRosenwasser's request, I've preserved some old behavior regarding the assignability of boolean-returning functions to type-predicate returning functions (ie, they're not assignable) which was likely to change so that allowing this can be saved for a later discussion.

**Comment by JsonFreeman** [comment]
@weswigham Thanks for the update. I have a few thoughts:

1. In terms of displaying the type, will quick info widen the type to boolean, or will it behave like type writer?
2. In terms of assignability of boolean-returning functions to type predicate-returning functions, I can see arguments for both, but I agree with @DanielRosenwasser's intuition about keeping it the same for now. It seems they probably should not be assignable, but then again, a boolean is assignable to a type predicate when it's not in return position.
3. You said "I propose not checking subtype assignability to the this type of this type guards (however, narrowing will still check if the type if a subtype to actually narrow)." What is the difference between the two checks you're referring to, in terms of the source type and target type for each?
4. This issue of vacuous truth and falsity does not seem unique to `this` type predicates, or even type predicates. It seems to apply for type guards across the board. I don't think it's a serious issue, but it's probably best to have a consistent policy for all type guards.
5. The widening approach we discussed seems like it would have a strange effect for function expressions and object literals. Would the following variables have different types?:
```ts
var v1: (x: any) => x is T;
var v2 = (x: any): x is T => true; // Does this widen?

var v3: { isT(): this is T; };
var v4 = { isT(): this is T { return true } }; // Does this widen?
```
It would seems unintuitive to me if v1 and v2 are different, or if v3 and v4 are different.

**Comment by weswigham** [comment]
@JsonFreeman 
1. Since it widens to boolean, in all meaningful situations (assignments, returns), it gets reinterpreted as a boolean.
2. And so it is, for now.
3. The check we do on type guards is that the RHS is a subtype of the LHS, meaning, in this case, that the RHS should be a subtype of the polymorphic `this` type - which is rather constraining, since this could mean that you can't guard from a base type to a sub-type. When narrowing, we check that they type guarded against is a subtype of the specific type being guarded on, and only narrow if this is the case (so you never narrow "across the lattice" as @DanielRosenwasser once put it).
4. We check, for parameter type guards, that the type of the parameter is a supertype of the type guarded against in the return type, hence why I stated that we should have not perform the equivalent for `this` guards, as it would be extremely limiting.
5. `v1` and `v2` are the same type, since we only widen on assignments/declarations - we never widen when the user has actually specified a predicate as the function's return type. `v3` and `v4` are both invalid because there is no polymorphic `this` available at those locations, but even if they were, there would be no widening in this location because, again, we don't widen return types within the signature.

**Comment by weswigham** [comment]
_By the way_, as @DanielRosenwasser  pointed out, the unification also makes this possible:
```ts
interface Array<T> {
  filter<U extends T>(f: (item: T, index: number, array: T[]) => item is U): U[];
}

var x: (string | number)[]

function isString(x: any): x is string {
    return typeof x === "string";
}

var y = x.filter(isString); // y is inferred as string[]
```

It's awesome.

File: src/compiler/binder.ts [hunk 10]
```diff
@@ -1275,6 +1276,16 @@ namespace ts {
             }
         }
 
+        function checkTypePredicate(node: TypePredicateNode) {
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: Consider just destructuring `parameterName` and `type`

File: src/compiler/types.ts [hunk 11]
```diff
@@ -2091,6 +2110,7 @@ namespace ts {
         ESSymbol                = 0x01000000,  // Type of symbol primitive introduced in ES6
         ThisType                = 0x02000000,  // This type
         ObjectLiteralPatternWithComputedProperties = 0x04000000,  // Object literal type implied by binding pattern has computed properties
+        PredicateType           = 0x08000000,
 
         /* @internal */
         Intrinsic = Any | String | Number | Boolean | ESSymbol | Void | Undefined | Null,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: More than anything else, can you leave a comment about the interaction here with `intrinsicName`?

File: src/compiler/diagnosticMessages.json [hunk 12]
```diff
@@ -1647,6 +1647,14 @@
         "category": "Error",
         "code": 2517
     },
+    "A 'this'-based type guard is not assignable to a parameter-based type guard": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: period, "assignable to or from" or "compatible with"

File: src/compiler/diagnosticMessages.json [hunk 13]
```diff
@@ -1647,6 +1647,14 @@
         "category": "Error",
         "code": 2517
     },
+    "A 'this'-based type guard is not assignable to a parameter-based type guard": {
+        "category": "Error",
+        "code": 2518
+    },
+    "A 'this'-based type predicate is only allowed in class or interface members, get accessors, or return type positions for functions and methods": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: "A 'this'-based type predicate is only allowed as the return type of functions, methods, and get accessors, or on members of classes and interfaces."

File: src/compiler/checker.ts [hunk 14]
```diff
@@ -14218,9 +14332,17 @@ namespace ts {
...
                 error(node, Diagnostics.A_type_predicate_is_only_allowed_in_return_type_position_for_functions_and_methods);
             }
+            else if (node.parameterName.kind === SyntaxKind.ThisType) {
+                if (!isInLegalThisTypePredicatePosition(node)) {
+                    error(node, Diagnostics.A_this_based_type_predicate_is_only_allowed_in_class_or_interface_members_get_accessors_or_return_type_positions_for_functions_and_methods);
+                }
+                else {
+                    getTypeFromThisTypeNode(node.parameterName as ThisTypeNode);
+                }
+            }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Is this intentionally supposed to return `void`?

File: src/compiler/checker.ts [hunk 15]
```diff
@@ -11017,9 +11130,14 @@ namespace ts {
 
             if (node.type) {
                 if (node.type.kind === SyntaxKind.TypePredicate) {
-                    const typePredicate = getSignatureFromDeclaration(node).typePredicate;
-                    const typePredicateNode = <TypePredicateNode>node.type;
-                    if (isInLegalTypePredicatePosition(typePredicateNode)) {
+                    const returnType = getReturnTypeOfSignature(getSignatureFromDeclaration(node));
+                    if (!returnType || !(returnType.flags & TypeFlags.PredicateType)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Consider using a `Debug.assert` here

File: src/compiler/checker.ts [hunk 16]
```diff
@@ -10999,6 +11099,19 @@ namespace ts {
             return false;
         }
 
+        function isInLegalThisTypePredicatePosition(node: Node): boolean {
+            if (isInLegalTypePredicatePosition(node)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I think this should be renamed to `isInLegalParameterTypePredicatePosition`

File: src/compiler/checker.ts [hunk 17]
```diff
@@ -6748,22 +6801,66 @@ namespace ts {
...
+                    const possibleIdentifier = skipParenthesizedNodes(accessExpression.expression);
+                    if (possibleIdentifier.kind === SyntaxKind.Identifier && getSymbolAtTypePredicatePosition(possibleIdentifier) === symbol) {
+                        return getNarrowedType(type, predicate.type, assumeTrue);
                     }
-                    return getNarrowedType(type, signature.typePredicate.type);
                 }
                 return type;
             }
 
+            function skipParenthesizedNodes(expression: Expression): Expression {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: There must be several functions that do exactly this. Check utilities here and in services

File: src/compiler/checker.ts [hunk 18]
```diff
@@ -5983,6 +6038,9 @@ namespace ts {
 
         function getWidenedType(type: Type): Type {
             if (type.flags & TypeFlags.RequiresWidening) {
+                if (type.flags & TypeFlags.PredicateType) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I think this is a less common case, so move it below the `undefined`/`null` check.

File: src/compiler/checker.ts (hunk 14 above)
Comment (weswigham) [inline src/compiler/checker.ts]: In this case, we're just using it for its error reporting (and caching of the type for later), rather then returning it immediately.

File: src/compiler/checker.ts (hunk 17 above)
Comment (weswigham) [inline src/compiler/checker.ts]: There aren't. :smile: 

There is another place in checker that uses an inline while loop though - I'll consolidate it.

File: src/compiler/checker.ts (hunk 18 above)
Comment (weswigham) [inline src/compiler/checker.ts]: :+1:

File: src/compiler/checker.ts (hunk 15 above)
Comment (weswigham) [inline src/compiler/checker.ts]: Can;t make it into an assertion since it can be `false` in error cases when a predicate type is found in an invalid position (for example, an index signature)

File: src/compiler/binder.ts (hunk 10 above)
Comment (weswigham) [inline src/compiler/binder.ts]: :+1:

File: tests/baselines/reference/typeGuardFunctionOfFormThisErrors.errors.txt [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        c;
+        holder;
+    }
+    
+    let detached = a.isFollower;
+    
+    if (detached()) {
+        a.follow();
+          ~~~~~~
+!!! error TS2339: Property 'follow' does not exist on type 'RoyalGuard'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/typeGuardFunctionOfFormThisErrors.errors.txt]: Awesome.

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by JsonFreeman** [comment]
Oh, so when you widen a function, you don't widen its declared return type.

I think my confusion about the "narrow across the lattice" point, is that I don't see the difference between the type being guarded, and the polymorphic `this` type (once instantiated). Are these in fact two different types?

**Comment by JsonFreeman** [comment]
And I agree that the type argument inference for type predicates is pretty nifty. Though I am not sure that the return type of a generic function at run time can realistically be based on the type predicate of a callback that's passed in.
